### PR TITLE
add nominatim geocoder

### LINKED ISSUE #185: Geocoding with Nominatim
Hello,

It would be nice to be able to use [Nominatim](http://wiki.openstreetmap.org/wiki/Nominatim) for geocoding. In my city, it is much more accurate than Google Maps.

### PR DESCRIPTION
**Pull request by PeterTheOne**
fix #185

### CODE REVIEW

File: includes/geocoders/Maps_NominatimGeocoder.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	 * @return string
+	 */	
+	protected function getRequestUrl( $address ) {
+		$urlArgs = [
+			'q' => urlencode( $address ),
+			'format' => 'jsonv2',
+			'limit' => 1,
+		];
+
+		return 'http://nominatim.openstreetmap.org/search?' . wfArrayToCgi($urlArgs);
```
Comment (JeroenDeDauw) [inline includes/geocoders/Maps_NominatimGeocoder.php]: No HTTPS?

**Comment by JeroenDeDauw** [comment]
Cool!

This is something you can easily write a test for. ... Or perhaps not entirely as `\Maps\Geocoder` is a pile of ****. Still, you could create a sane geocoding class and simply have an adapter deriving from `\Maps\Geocoder` that uses that.

```php
class Maps\Geocoders\NominatimGeocoder {
    function __construct( SomeWebRequestThing $internets ) { $this->internets = $internets; }
    function geocode( string $address ): LatLongValue {
       // weee
       $this->internets->request( /* ... */ );
       // weee
    }
}
```

Perhaps implementing some new `Maps\Geocoders\NonSuckyGeocoder` interface.

You can then have a unit test that verifies the URLs get build correctly given the input, and one that verifies various kinds of API responses get turned into the function return value. And to really test all the things, you could add a system test that runs some requests against the real API (this test would not be run by default, only with specific flag).

**Comment by PeterTheOne** [comment]
I didn't feel like rewriting `\Maps\Geocoder` and I'm not really sure how to do better.

**Comment by JeroenDeDauw** [comment]
My suggestion here does not involve rewriting anything already there. It describes how you can add the new functionality in such a way that it does not add to the pile of legacy crud already there.

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Maps\Test;
+
+/**
+ * @covers MapsNominatimGeocoder
+ *
+ * @since 4.0
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: These make no sense for non-package-public classes IMO

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @group Maps
+ *
+ * @licence GNU GPL v2+
+ * @author Peter Grassberger < petertheone@gmail.com >
+ */
+class NominatimGeocoderTest extends \PHPUnit_Framework_TestCase {
+
+	protected static function getMethod( $name ) {
+		$class = new \ReflectionClass( 'MapsNominatimGeocoder' );
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: Such reflection!

`MapsNominatimGeocoder::class`

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		];
+	}
+
+	/**
+	 * @covers MapsNominatimGeocoder::getRequestUrl
+	 *
+	 * @dataProvider addressProvider
+	 */
+	public function testGetRequestUrl( $address, $expected ) {
+		$getRequestUrl = self::getMethod('getRequestUrl');
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: spacing

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		$geocoder = new \MapsNominatimGeocoder( 'nominatim' );
+		$actual = $getRequestUrl->invokeArgs($geocoder, array( $address ));
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: spacing

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		];
+	}
+
+	/**
+	 * @covers MapsNominatimGeocoder::parseResponse
+	 *
+	 * @dataProvider responseProvider
+	 */
+	public function testParseResponse( $response, $expected ) {
+		$parseResponse = self::getMethod('parseResponse');
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: spacing

File: tests/phpunit/geocoders/NominatimGeocoderTest.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		$geocoder = new \MapsNominatimGeocoder( 'nominatim' );
+		$actual = $parseResponse->invokeArgs($geocoder, array( $response ));
```
Comment (JeroenDeDauw) [inline tests/phpunit/geocoders/NominatimGeocoderTest.php]: spacing

**Comment by JeroenDeDauw** [comment]
If you do not understand the approach I'm suggesting and think it would be instructive if I did it then I can give that a go.

**Comment by PeterTheOne** [comment]
> If you do not understand the approach I'm suggesting and think it would be instructive if I did it then I can give that a go.

I think I know where you are going but I don't get why you would want to create an extra class and interface to wrap one new geocoder so we can test it if you say the Geocoder class is the problem. Why not restructure that one? In turn maybe I don't get it. :P If you want to do it, please do, I also think it could be instructive.

**Comment by JeroenDeDauw** [comment]
Thanks. I'll submit a follow up

**Comment by JeroenDeDauw** [comment]
Will have to be for after my holiday tho...

**Comment by JeroenDeDauw** [comment]
Follow up at https://github.com/JeroenDeDauw/Maps/pull/214

**Pull request merged**